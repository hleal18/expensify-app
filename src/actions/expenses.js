import uuid from 'uuid';
import database from '../firebase/firebase';

//ADD_EXPENSE
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//ASYNC REDUX ACTIONS
//Forma asincronica de retornar el objeto para posteriormente
//el dispatch lo aproveche.
export const startAddExpense = (expenseData = {}) => {
    //Retorna una funcion.
    //Se busca que el componente primero realice operaciones con la BD.
    //Como son asíncrónicas, esto permite que al terminar aquella tarea,
    //se invoque el dispatch para así mantener todos los datos correctamente
    //actualizados y visualizados.
    //Para guardar expenses a un usuario especifico, se usa el segundo parametro.
    //que es getState que contiene el uid del user.
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0 
        } = expenseData;
        const ref = database.ref(`users/${uid}/expenses`).push();
        const key = ref.key;
        const expense = { description, note, amount, createdAt };
        dispatch(addExpense({id: key, ...expense}));
        dispatch(changeStateExpense(key, 'Adding new expense'));
        //Se retorna un promise cuando se complete exitosamente la adicion de un expense
        //de acuerdo al uid del usuario.
        return ref.set(expense).then((ref) => {
            dispatch(changeStateExpense(key, 'Saved expense'));
        });
    };
};

//REMOVE_EXPENSE
export const removeExpense = ({ id } = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//FUNCION ASINCRONICA REMOVE_EXPENSE
export const startRemoveExpense = ({ id }) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        dispatch(changeStateExpense(id, 'Removing expense'))
        return database.ref(`users/${uid}/expenses/${id}`).remove().then(() => {
            dispatch(removeExpense({ id }));
        });
    };    
};

//EDIT_EXPENSE
export const editExpense = (id, updates) => ({
    type: 'EDIT_EXPENSE',
    id,
    updates    
});

//FUNCION ASINCRONICA EDIT_EXPENSE
export const startEditExpense = (id, update) => {
    return (dispatch, getState) => {
        dispatch(editExpense(id, update));
        dispatch(changeStateExpense(id, 'Editing expense'));
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/expenses/${id}`).update(update).then(() => {
            dispatch(changeStateExpense(id, 'Saved expense'));
        });
    };
};

//MANIPULA EL REDUX STORE, OBTIENE LOS EXPENSES
export const setExpenses = (expenses) => ({
    type: 'SET_EXPENSES',
    expenses
});

//FORMA ASÍNCRONICA DE SET EXPENSES
export const startSetExpenses = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        //Se obtienen los expenses de acuerdo al uid contenida en el state.
        return database.ref(`users/${uid}/expenses`)
            .once('value')
            .then((snapshot) => {
                const expenses = [];

                snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id: childSnapshot.key,
                        ...childSnapshot.val()
                    });
                });
                dispatch(setExpenses(expenses));
                expenses.forEach((expense) => {dispatch(changeStateExpense(expense.id, 'Saved expense'))});
            });
    };
};

//ACCION PARA ESTABLECER EL ESTADO DE UN EXPENSE
export const changeStateExpense = (id, message) => ({
    type: 'CHANGE_STATE_MESSAGE',
    id,
    message
});