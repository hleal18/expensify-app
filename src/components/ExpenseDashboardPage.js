import React from 'react';

const ExpenseDashboardPage = () => (
    <div>
        This is from my dashboard componente
    </div>
);

export default ExpenseDashboardPage;