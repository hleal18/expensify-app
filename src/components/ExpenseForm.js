import React, { Component } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';

// Se recibe una instancia de moment.
// tiene el momento actual.
// const now = moment();
// imprime el objeto
// console.log(now);
// imprime la fecha en formato
// console.log(now.format());
// imprime el mes recortado.
// console.log(now.format('MMM'));
// imprime el mes recortado con el dia del mes.
// console.log(now.format('MMM Do'));
// imprime el mes recortado con el dia del mes y el año en cuatro digitos.
// console.log(now.format('MMM Do, YYYY'));


//Se busca aprovechar el estado interno del componente de los datos de los inputs
//y cuando se haya finalizado, esos datos del estado, se guardarán en redux mediante
//el nuevo expense.
class ExpenseForm extends Component {
    constructor(props){
        super(props);
        this.state = {
            description: props.expense ? props.expense.description : '',
            amount: props.expense ? (props.expense.amount / 100).toString() : '',
            note: props.expense ? props.expense.note : '',
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            calendarFocused: false,
            error: ''
        };
    }
    
    //Se enlaza la descripción del estado con la que escribe el usuario.
    //se mantiene guardado el estado en todo momento.
    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState(() => ({ description }));
    };
    onNoteChange = (e) => {
        const note = e.target.value;
        this.setState(() => ({ note }));
    };
    //Se busca que para el input de amount, se acepten numeros de hasta dos decimales
    //nada más.
    onAmountChange = (e) => {
        const amount = e.target.value;
        //Se usa una expresión regular. (regular expressiones)
        //es un tema grande a consultar.
        //Se busca la documentación en regex101.com
        //se fuerza que haya por lo menos un numero antes del punto decimal.
        //tambien se permite que el usuario borre toda la cantidad.
        if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
            this.setState(() => ({ amount }));
        }
    };
    onDateChange = (createdAt) => {
        if(createdAt) {
            this.setState(() => ({ createdAt }));
        }
    };
    onFocusChange = ({ focused }) => {
        this.setState(() => ({ calendarFocused: focused }))
    };
    onSubmit = (e) => {
        e.preventDefault();

        if(!this.state.description || !this.state.amount) {
            //Set error state equal to 'Please provide description and amount.'
            this.setState(() => ({ error: 'Please provide description and amount.' }));
        } else {
            //Clear the error
            this.setState(() => ({ error: '' }));
            this.props.onSubmit({
                description: this.state.description,
                amount: parseFloat(this.state.amount),
                note: this.state.note,
                //Se le pasa el timestamp. No el objeto moment.
                createdAt: this.state.createdAt.valueOf()
            });
        }
    };
    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Description"
                        autoFocus
                        value={this.state.description}
                        onChange={this.onDescriptionChange}
                    />
                    <input
                        type="text"
                        placeholder="Amount"
                        value={this.state.amount}
                        onChange={this.onAmountChange}
                    />
                    <SingleDatePicker
                        date={this.state.createdAt}
                        onDateChange={this.onDateChange}
                        focused={this.state.calendarFocused}
                        onFocusChange={this.onFocusChange}
                        numberOfMonths={1}
                        isOutsideRange={() => false}
                    />
                    <textarea
                        placeholder="Add a note for your expense (optional)"
                        value={this.state.note}
                        onChange={this.onNoteChange}
                    >
                    </textarea>
                    <button>Add expense</button>
                </form>
            </div>
        );
    }
}

//Una vez terminada la implementación del componente y sus estados queda incluir la conexión
//con los datos de redux.
//Debido a que este componente será reutilizado, no combiene usarlo aquí sino en los componentes
//que se beneficiarán de éste.

export default ExpenseForm;
