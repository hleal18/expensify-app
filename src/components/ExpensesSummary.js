import React from 'react';
import { connect } from 'react-redux';
import selectExpensesTotal from '../selectors/expense-total';
import getVisibleExpenses from '../selectors/expenses';
import numeral from 'numeral';

export const ExpensesSummary = (props) => {
    const expenseWord = props.expenseCount === 1 ? 'expense' : 'expenses';
    const formattedExpenseTotal = numeral(props.expensesTotal/100).format('$0,0.00');
    return (
        <div>
            <h1>Viewing {props.expenseCount} {expenseWord} totalling {formattedExpenseTotal}</h1>
        </div>
    );
};

const mapStateToProps = (state) => {
    const visibleExpenses = getVisibleExpenses(state.expenses, state.filters);

    return {
        expenseCount: visibleExpenses.length,
        expensesTotal: selectExpensesTotal(visibleExpenses)
    };
};

export default connect(mapStateToProps)(ExpensesSummary);