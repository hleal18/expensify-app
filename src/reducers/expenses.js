const expensesReducerDefaultState = [];

export default (state = expensesReducerDefaultState, action) => {
    switch (action.type) {
        case 'ADD_EXPENSE':
            //Se puede usar el es6 spread operator en arreglos.
            return [
                ...state,
                action.expense
            ];            
        case 'REMOVE_EXPENSE':
            return state.filter(({ id }) => id !== action.id);            
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if (expense.id === action.id) {
                    //Se ponen los atributos que ya posee expense
                    //Se sobreescriben los que poseen updates del Action generator.
                    return {
                        ...expense,
                        ...action.updates
                    };
                } else {
                    return expense;
                }
            });            
        default:
            return state;
    }
};