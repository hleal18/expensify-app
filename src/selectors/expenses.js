import moment from 'moment';
//milliseconds are counting in milliseconds.
// 0 = January 1st 1970 (unix epoch) at midnight

//Get visible expenses
export default (expenses, { text, sortBy, startDate, endDate }) => {
    return expenses.filter((expense) => {
        ///Se obtiene una instancia del timestamp del expense evaluado.
        const createdAtMoment = moment(expense.createdAt);
        //Se compara para el comienzo y el final del filtro elegido mediante
        //isSameOfBefore con respecto al día.
        const startDateMatch = startDate ? startDate.isSameOrBefore(createdAtMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(createdAtMoment, 'day') : true;
        //Se encarga de encontrar una coincidencia de el texto en la descripción.
        const textMatch = expense.description.toLowerCase().includes(text.toLowerCase());

        return startDateMatch && endDateMatch && textMatch;
    }).sort((a, b) => {
        if (sortBy === 'date') {
            return a.createdAt < b.createdAt ? 1 : -1;
        } else if (sortBy == 'amount') {
            return a.amount < b.amount ? 1 : -1;
        }
    });
};