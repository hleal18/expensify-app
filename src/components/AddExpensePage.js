import React from 'react';
import ExpenseForm from './ExpenseForm';
import { connect } from 'react-redux';
import { addExpense } from '../actions/expenses';

//Se reutiliza ExpenseForm y se conecta con connect de redux
//Se pasa a una clase
//Como se quiere testear es la versión aislada de AddExpensePage
//Se exporta individualmente (no como default)
export class AddExpensePage extends React.Component {
    onSubmit = (expense) => {
        //Debido a la dificultad de testear lo siguiente.
        //Ya que se invoca una función y dentro de ella hay otra.
        //Se implementa el mapDispatchToProps.
        //props.dispatch(addExpense(expense));
        //Se agrega el prop que está enlazado con la nueva
        //funcion.
        this.props.onSubmit(expense);
        //Lo provee router. Permite redireccionar cuando se suban los datos.
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
            <h1>Add expense</h1>
            <ExpenseForm
                onSubmit={this.onSubmit}
            />
            </div>
        );
    }
}

//Se agrega el mapDispatchToProps
//Se ponen las propiedades que invocarán a dispatch.
//Establece la invocación del dispatch, ahora accesible
//desde las props del componente.
const mapDispatchToProps = (dispatch) => ({
    onSubmit: (expense) => dispatch(addExpense(expense))
});

export default connect(undefined, mapDispatchToProps)(AddExpensePage);