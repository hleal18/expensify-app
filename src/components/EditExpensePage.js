import React from 'react';
import { connect } from 'react-redux';
import ExpenseForm from './ExpenseForm';
import { editExpense, removeExpense } from '../actions/expenses';

export class EditExpensePage extends React.Component {
    onSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id, expense);
        this.props.history.push('/');
    }
    onClick = () => {
        this.props.removeExpense({ id: this.props.expense.id });
        this.props.history.push('/');
    }
    render() {
        return (
            <div>
                {/*Se accede al id que es pasado por el router y accedido mediante
            match y params. Se encuentra en la url.
            Consultar en la consola el history en la herramienta React.
        */}
                <ExpenseForm
                    expense={this.props.expense}
                    onSubmit={this.onSubmit}
                />
                <button
                    onClick={this.onClick}
                >
                    Remove
            </button>
            </div>
        )
    }
}

//Se debe acceder a los props del componente original, con ello es suficiente
//agregar el segundo argumento de props.
const mapStateToProps = (state, props) => {
    return {
        expense: state.expenses.find((expense) => (expense.id === props.match.params.id))
    }
}

const mapDispathToProps = (dispatch) => ({
    editExpense: (id, expense) => (dispatch(editExpense(id, expense))),
    removeExpense: (data) => (dispatch(removeExpense(data)))
});

export default connect(mapStateToProps, mapDispathToProps)(EditExpensePage);