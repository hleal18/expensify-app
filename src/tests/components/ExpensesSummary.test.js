import { shallow } from 'enzyme';
import {ExpensesSummary} from '../../components/ExpensesSummary';
import React from 'react';
import expenses from '../fixtures/expenses';
import getExpensesTotal from '../../selectors/expense-total';

test('should render correctly ExpensesSummary for one expense', () => {
    const expensesTest = [expenses[0]];
    const wrapper = shallow(
        <ExpensesSummary
            expenseCount={expensesTest.length}
            expensesTotal={getExpensesTotal(expensesTest)}
        />
    );
    expect(wrapper).toMatchSnapshot();
});

test('should render correctly ExpensesSummary for multiple expenses', () => {
    const wrapper = shallow(
        <ExpensesSummary 
        expenseCount={expenses.length} 
        expensesTotal={getExpensesTotal(expenses)} 
        />
    );
    expect(wrapper).toMatchSnapshot();
});