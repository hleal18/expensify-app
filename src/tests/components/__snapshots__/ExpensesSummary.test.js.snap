// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly ExpensesSummary for multiple expenses 1`] = `
<div>
  <h1>
    Viewing 
    3
     
    expenses
     totalling 
    $1,141.95
  </h1>
</div>
`;

exports[`should render correctly ExpensesSummary for one expense 1`] = `
<div>
  <h1>
    Viewing 
    1
     
    expense
     totalling 
    $1.95
  </h1>
</div>
`;
